### PR TITLE
Added adaptive lr to training loop.

### PR DESCRIPTION
**Pull request by brudfors**
I have added adaptive lr to the training loop, a quick test (2D) shows slightly improved performance for segmentation:

with adapt-lr
k=BG | avg(dice)=0.9777, N=8
k=GM | avg(dice)=0.8718, N=8
k=WM | avg(dice)=0.9170, N=8
k=CSF | avg(dice)=0.7978, N=8

without adapt-lr
k=BG | avg(dice)=0.9752, N=8
k=GM | avg(dice)=0.8595, N=8
k=WM | avg(dice)=0.9113, N=8
k=CSF | avg(dice)=0.7257, N=8

### CODE REVIEW

**Comment by balbasty** [comment]
Just a quick warning: you also have the augmentation in there. We’ll have to make sure there are no conflicts when we merge the final versions

File: nitorch/nn/training/_training.py [hunk 1]
```diff
@@ -582,7 +598,7 @@ def init(self):
             self.save_random_state()
             self.epoch = self.initial_epoch
             self.model.to(dtype=self.dtype, device=self.device)
-            self._eval(self.epoch)
+            _ = self._eval(self.epoch)
```
Comment (balbasty) [inline nitorch/nn/training/_training.py:601]: Do we need that? I thought we could call functions without return values even if the do return one

File: nitorch/nn/training/_training.py [hunk 2]
```diff
@@ -563,15 +574,20 @@ def train(self):
                 self._eval(self.epoch)
                 self._save(self.epoch)
                 for self.epoch in range(self.epoch+1, self.nb_epoch+1):
-                    self._train(self.epoch)
-                    self._eval(self.epoch)
+                    _ = self._train(self.epoch)
+                    val_loss = self._eval(self.epoch)
```
Comment (balbasty) [inline nitorch/nn/training/_training.py:578]: There is not always a validation set! I don’t know if you should give it the training loss, or the eval loss if it exists else the training loss?

File: nitorch/nn/training/_training.py [hunk 3]
```diff
@@ -563,15 +574,20 @@ def train(self):
...
                     self._save(self.epoch)
-
+                    # incorporate learning rate scheduler
+                    if self.scheduler is not None:
+                        if isinstance(self.scheduler, ReduceLROnPlateau):
```
Comment (balbasty) [inline nitorch/nn/training/_training.py:582]: I wonder if it is possible to inspect the scheduler (using the built in  `inspect` package) to know if a loss must be provided or not. That would be more generic.

**Review by balbasty — COMMENTED** [review]
thanks !

The main issue is the one related to the eval loss that doesn’t always exist.
(I didn’t check but I have never tried the trainer class without giving it an eval set so maybe it breaks anyway... but it would be a bug and shouldn’t)

**Comment by brudfors** [comment]
> Just a quick warning: you also have the augmentation in there. We’ll have to make sure there are no conflicts when we merge the final versions

Omg, my misstake, I think I have fixed it.

File: nitorch/nn/training/_training.py (hunk 1 above)
Comment (brudfors) [reply nitorch/nn/training/_training.py:601]: Removed.

File: nitorch/nn/training/_training.py (hunk 2 above)
Comment (brudfors) [reply nitorch/nn/training/_training.py:578]: Thanks for catching that, fixed.

File: nitorch/nn/training/_training.py (hunk 3 above)
Comment (brudfors) [reply nitorch/nn/training/_training.py:582]: Looking at the implemented scheds in torch.optim.lr_scheduler, ReduceLROnPlateau is the only one that takes an input to its step call. So maybe it is okay to keep it like this for now?
